lint/problem.go

```diff
@@ -15,34 +15,128 @@
 package lint
 
 import (
+	"encoding/json"
+
+	dpb "github.com/golang/protobuf/protoc-gen-go/descriptor"
 	"github.com/jhump/protoreflect/desc"
+	"gopkg.in/yaml.v2"
 )
 
 // Problem contains information about a result produced by an API Linter.
 type Problem struct {
 	// Message provides a short description of the problem.
-	Message string `json:"message" yaml:"message"`
+	Message string
 
 	// Suggestion provides a suggested fix, if applicable.
-	Suggestion string `json:"suggestion,omitempty" yaml:"suggestion,omitempty"`
+	Suggestion string
 
 	// Descriptor provides the descriptor related
 	// to the problem. If present and `Location` is not
 	// specified, then the starting location of the descriptor
 	// is used as the location of the problem.
-	Descriptor desc.Descriptor `json:"-" yaml:"-"`
+	Descriptor desc.Descriptor
 
 	// Location provides the location of the problem.
 	// DO NOT SET: Set the descriptor instead.
-	Location Location `json:"location" yaml:"location"`
+	Location *dpb.SourceCodeInfo_Location
 
 	// RuleID provides the ID of the rule that this problem belongs to.
 	// DO NOT SET: this field will be set by the linter based on rule info
 	// and user configs.
-	RuleID RuleName `json:"rule_id" yaml:"rule_id"`
+	RuleID RuleName
 
 	// DO NOT SET:  this field will be set by the linter based on user configs.
-	Category string `json:"category,omitempty" yaml:"category,omitempty"`
+	Category string
 
 	noPositional struct{}
 }
+
+// MarshalJSON defines how to represent a Problem in JSON.
+func (p Problem) MarshalJSON() ([]byte, error) {
+	return json.Marshal(p.marshal())
+}
+
+// MarshalYAML defines how to represent a Problem in YAML.
+func (p Problem) MarshalYAML() ([]byte, error) {
+	return yaml.Marshal(p.marshal())
+}
+
+// Marshal defines how to represent a serialized Problem.
+func (p Problem) marshal() interface{} {
```

Comment:
:+1: much happier with this than previous solution. Now we can also make `ruleId` and `category` private, rather than asking developers not to set them in comments.